src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -206,5 +290,120 @@ public Task UnloadAsync()
+                {
+                    return Task.CompletedTask;
+                }
+
+                _whenNominatedTask ??= new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            }
+
+            var continuationTask = _whenNominatedTask.Task.ContinueWith(t =>
+            {
+                lock (SyncObject)
+                {
+                    if (t.IsFaulted)
+                    {
+                        _whenNominatedTask.SetException(t.Exception);
+                    }
+                    else
+                    {
+                        _whenNominatedTask.TrySetCanceled();
+                    }
+                }
+            }, cancellationToken,
+            TaskContinuationOptions.OnlyOnFaulted,
+            TaskScheduler.Default);
+
+            return continuationTask;
+        }
+
+        // True means the project system plans to call NominateProjectAsync in the future.
+        bool IVsProjectRestoreInfoSource.HasPendingNomination
+        {
+            get => CheckIfHasPendingNomination();
+        }
+
+        private bool CheckIfHasPendingNomination()
+        {
+            lock (SyncObject)
+            {
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider);
+                Assumes.Present(_project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject);
+
+                if (SourceBlock.Completion.IsFaulted)
+                {
+                    return false;
+                }
+
+                // Avoid possible deadlock.
+                // Because RestoreCoreAsync() is called inside a dataflow block it will not be called with new data
+                // until the old task finishes. So, if the project gets nominating restore, it will not get updated data.
+                if (IsNugetRestoreFinished())
+                {
+                    return false;
+                }
+
+                ConfiguredProject? activeConfiguredProject = _project.Services.ActiveConfiguredProjectProvider.ActiveConfiguredProject;
+
+                // Nuget should wait until the project at least nominates once.
+                if (!_savedNominatedConfiguredVersion.ContainsKey(activeConfiguredProject.ProjectConfiguration))
+                {
+                    return true;
+                }
+
+                // Nuget should not wait for projects that failed DTB
+                if (SourceBlock.Completion.IsCompleted)
+                {
+                    return false;
+                }
+
+                // After the first nomination, we should check the saved nominated version
+                return IsSavedNominationEmptyOrOlder(activeConfiguredProject);
+            }
+        }
+
+        private bool IsNugetRestoreFinished()
+        {
+            // TODO : if NominateForRestoreAsync() has not finished, return false
+            return !_nugetRestoreStarted;
```

Comment:
@lifengl 
This flag is set to true just before `NominateForRestoreAsync()` is executed and it is set to false after `NominateForRestoreAsync()` finish